### PR TITLE
Replace `#[default_method_body_is_const]` with `#[const_trait]`

### PR DESCRIPTION
**Pull request by oli-obk**
pulled out of #96077

related issues:  #67792 and #92158

cc @fee1-dead 

This is groundwork to only allowing `impl const Trait` for traits that are marked with `#[const_trait]`. This is necessary to prevent adding a new default method from becoming a breaking change (as it could be a non-const fn).

### CODE REVIEW

File: compiler/rustc_passes/src/check_attr.rs [hunk 1]
```diff
@@ -2185,6 +2174,8 @@ impl CheckAttrVisitor<'_> {
                 "attribute `{}` without any lints has no effect",
                 attr.name_or_empty()
             )
+        } else if attr.name_or_empty() == sym::default_method_body_is_const {
+            format!("`default_method_body_is_const` has been removed")
```
Comment (fee1-dead) [inline compiler/rustc_passes/src/check_attr.rs:2178]:
```suggestion
            format!("`default_method_body_is_const` has been replaced with `#[const_trait]` on traits")
```

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 2]
```diff
@@ -890,9 +890,11 @@ fn should_encode_mir(tcx: TyCtxt<'_>, def_id: LocalDefId) -> (bool, bool) {
             let needs_inline = (generics.requires_monomorphization(tcx)
                 || tcx.codegen_fn_attrs(def_id).requests_inline())
                 && tcx.sess.opts.output_types.should_codegen();
-            // The function has a `const` modifier or is annotated with `default_method_body_is_const`.
-            let is_const_fn = tcx.is_const_fn_raw(def_id.to_def_id())
-                || tcx.has_attr(def_id.to_def_id(), sym::default_method_body_is_const);
+            // The function has a `const` modifier or is in a `#[const_trait]`.
+            let mut is_const_fn = tcx.is_const_fn_raw(def_id.to_def_id());
+            if let Some(trait_) = tcx.trait_of_item(def_id.to_def_id()) {
+                is_const_fn = is_const_fn || tcx.has_attr(trait_, sym::const_trait);
```
Comment (jhpratt) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:896]:
```suggestion
                is_const_fn ||= tcx.has_attr(trait_, sym::const_trait);
```

(I wish we had clippy enabled for this repository)

File: compiler/rustc_passes/src/check_attr.rs (hunk 1 above)
Comment (jhpratt) [reply compiler/rustc_passes/src/check_attr.rs:2184]: This was marked resolved but was not changed.

**Review by jhpratt — COMMENTED** [review]
Part of me would like the syntax to require `const fn` for the default methods, as it's not immediately clear that it's the case with this change. But I understand that the syntax/attributes is largely in flux, so that's not a deal-breaker for me.

Overall, the PR looks good. Just some minor nits.

File: compiler/rustc_passes/src/check_attr.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_passes/src/check_attr.rs]: Oops thanks! I did apply it and screw up a rebase

**Comment by jhpratt** [comment]
Huh, apparently that's not a thing in Rust? Even [a minimal example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=722063f4965f8e637becfd1cbb409452) reproduces that error message.

**Comment by fee1-dead** [comment]
I'd think that for `a ||= b`, whether `b` gets evaluated depends on `a`, which is pretty confusing to me.

So for `a ||= foo(42)`, `foo` could be not called if `a` is `true`.

**Comment by jhpratt** [comment]
Yeah, I brought it up on Zulip. I suggested a diagnostic improvement because it didn't sound like something people wanted.

**Review by jhpratt — APPROVED** [review]
LGTM pending rebase.

File: compiler/rustc_const_eval/src/transform/check_consts/check.rs [hunk 3]
```diff
@@ -774,13 +772,12 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
...
-                            // #[default_method_body_is_const], and the callee is in the same
-                            // trait.
-                            let callee_trait = tcx.trait_of_item(callee);
-                            if callee_trait.is_some()
-                                && tcx.has_attr(caller.to_def_id(), sym::default_method_body_is_const)
-                                && callee_trait == tcx.trait_of_item(caller)
+                            // At this point, it is only legal when the caller is in a trait
+                            // marked with #[const_trait], and the callee is in the same trait.
+                            let mut nonconst_call_permission = false;
+                            if let Some(callee_trait) = tcx.trait_of_item(callee)
```
Comment (compiler-errors) [inline compiler/rustc_const_eval/src/transform/check_consts/check.rs:778]: I wish `if` chains could be booleans, sigh

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 4]
```diff
@@ -891,9 +891,11 @@ fn should_encode_mir(tcx: TyCtxt<'_>, def_id: LocalDefId) -> (bool, bool) {
             let needs_inline = (generics.requires_monomorphization(tcx)
                 || tcx.codegen_fn_attrs(def_id).requests_inline())
                 && tcx.sess.opts.output_types.should_codegen();
-            // The function has a `const` modifier or is annotated with `default_method_body_is_const`.
-            let is_const_fn = tcx.is_const_fn_raw(def_id.to_def_id())
-                || tcx.has_attr(def_id.to_def_id(), sym::default_method_body_is_const);
+            // The function has a `const` modifier or is in a `#[const_trait]`.
+            let mut is_const_fn = tcx.is_const_fn_raw(def_id.to_def_id());
+            if let Some(trait_) = tcx.trait_of_item(def_id.to_def_id()) {
```
Comment (compiler-errors) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:896]: could use `is_const_default_method` but rename it maybe to `is_const_default_trait_item`? Maybe not.

File: compiler/rustc_span/src/symbol.rs [hunk 5]
```diff
@@ -504,6 +504,8 @@ symbols! {
         const_raw_ptr_deref,
         const_raw_ptr_to_usize_cast,
         const_refs_to_cell,
+        const_slice_ptr,
```
Comment (compiler-errors) [inline compiler/rustc_span/src/symbol.rs:507]: where is this symbol used?

**Review by compiler-errors — APPROVED** [review]
r=me, left a few comments (or ignore them, that's fine too)

File: compiler/rustc_const_eval/src/transform/check_consts/check.rs (hunk 3 above)
Comment (fee1-dead) [reply compiler/rustc_const_eval/src/transform/check_consts/check.rs:778]: I'm not sure what this means :)

Comment (compiler-errors) [reply compiler/rustc_const_eval/src/transform/check_consts/check.rs:778]: sorry, meant `let` chains. just don't like how this has to be a mutable assignment lol. not an actionable comment.

**Comment by pnkfelix** [comment]
* The primary regressions are mostly in variants of stm32f4, with a two serde and one syn thrown in for good measure.
* The improvements are solely to ctfe stress test, which I guess makes sense given the PR?
* Anyway, the regressions seem minor, and they are contained to an unstable feature that the stdlib is using.
* Marking as triaged.

@rustbot label: +perf-regression-triaged

File: compiler/rustc_passes/src/check_const.rs [hunk 6]
```diff
@@ -64,66 +63,6 @@ pub(crate) fn provide(providers: &mut Providers) {
...
-                tcx
-                    .sess
-                    .struct_span_err(
-                        item.span,
-                        "const trait implementations may not use non-const default functions",
-                    )
-                    .note(&format!("`{}` not implemented", not_implemented))
-                    .emit();
-            }
-        }
```
Comment (fee1-dead) [inline compiler/rustc_passes/src/check_const.rs:123]: I think we forgot to actually forbid `impl const` on non `#[const_trait]` which would be here

**Pull request merged**